### ISSUE TITLE
Add ability to skip rows with invalid/unavailable CLR class

### ISSUE THREAD

**Issue by miyconst** [Starcounter/StarDump.Migrator.2.2#13]
Consider the following database structure.

```cs
[Database]
public class Parent {
}

[Database]
public class Child : Parent {
}
```

If there are some `Child` row, but the `Child` class was deleted, then the following query fails.

```sql
SELECT *
FROM Parent
```

The `Child` rows are also available through the `Parent` class, but there is no CLR class for it.

It should be possible to rule out such situations and skip the rows in the `StarDump.Migrator.2.2` tool.

**Comment by miyconst** [comment]
I tried to use `Enumerator` instead of `foreach` loop and `try/catch` the retrieval of `Current` object to be able to skip the error rows, but it fails on `enumerator.MoveNext()`, so not an option.

Also tried to use SQL `FETCH` and `OFFSET`, same situation.

There might be several solutions of how to achieve desired behavior:

1. Apply a hack/fix in Level1 which would allow to skip such rows. @k-rus, @alemoi 
2. Generate and load CLR classes during run time. @per-samuelsson 

@dan31 this issue does not seem to be a trivial fix for 1-2 hours of work, still it is a showstopper for 2.2 -> 2.3 migration. How do we proceed?

**Comment by dan31** [comment]
@miyconst in which particular situation this has appeared? Need details. Do you mean there is some situation where the state of 2.2-run database is, let's say, screwed up with some schema migrations?

**Comment by per-samuelsson** [comment]
@miyconst 

Since this is exclusively an issue for 2.2, it seem like a hack (if possible) would be a better option that runtime code generation. How many customers are actually still on 2.2, with real data?

**Comment by per-samuelsson** [comment]
Out of curiosity, would this query work:

```sql
SELECT Name
FROM Parent
```

(with Name part of Parent, and Child not loaded)

**Comment by miyconst** [comment]
This situation has happened during 2.2 -> 2.3 migration process for @malx122. In his case the `Child` table was deleted long time ago, but it's records are still in the database.

**Comment by miyconst** [comment]
>Out of curiosity, would this query work: ...

No. Even `SELECT COUNT(*)` does not work.

**Comment by miyconst** [comment]
>Do you mean there is some situation where the state of 2.2-run database is, let's say, screwed up with some schema migrations?

Yes. See my comment above.

**Comment by miyconst** [comment]
>Since this is exclusively an issue for 2.2, it seem like a hack (if possible) would be a better option that runtime code generation. How many customers are actually still on 2.2, with real data?

Currently - both, who is migrating from 2.2 to 2.3 (Tunity & AdFenix). Till now I was able to find manual workarounds. The latest database @malx122 is trying to unload requires too much manual work. That is why this issue is here.

I would prefer it be implemented on the `StarDump.Migrator.2.2` app level, to don't ask customers to install some special edition of Starcounter 2.2 just for the sake of unload.

**Comment by per-samuelsson** [comment]
>  In his case the Child table was deleted long time ago, but it's records are still in the database.

@dan31, we should have had that `vacuum()` thingy already in 2.2! 😄

**Comment by per-samuelsson** [comment]
> I would prefer it be implemented on the `StarDump.Migrator.2.2` app level, to don't ask customers to install some special edition of Starcounter 2.2 just for the sake of unload.

Yeah, totally agree. Lets see if @k-rus can give some hints.

**Comment by k-rus** [comment]
> In his case the Child table was deleted long time ago, but it's records are still in the database.
> @dan31, we should have had that vacuum() thingy already in 2.2!

The implementation wise it will be totally different. Vacuum name is more appropriate for 2.3 and above, while in 2.2. it is just drop with delete.



> I would prefer it be implemented on the StarDump.Migrator.2.2 app level, to don't ask customers to install some special edition of Starcounter 2.2 just for the sake of unload.
> Yeah, totally agree. Lets see if @k-rus can give some hints.

Since MoveNext fails, it means that CLR class instantiation is somewhere deep inside. I haven't looked in the code. We might be lucking to make fix or hack in a short amount of time, or it might require to do really ugly hack, which will break part of existing functionality. To investigate this is 1-2 days of work I guess.

**Comment by bigwad** [comment]
Can Migrator just skip the exception?

As to the error itself - it's not specific for 2.2. I expect 2.3 and above to be the subject as well.

**Comment by per-samuelsson** [comment]
> Can Migrator just skip the exception?

See:

> I tried to use `Enumerator` instead of `foreach` loop and `try/catch` the retrieval of `Current` object to be able to skip the error rows, but it fails on `enumerator.MoveNext()`, so not an option.

**Comment by bigwad** [comment]
Well, what if you keep calling MoveNext()? Will it finally pass beyond invalid objects?

**Comment by per-samuelsson** [comment]
> As to the error itself - it's not specific for 2.2. I expect 2.3 and above to be the subject as well.

The difference is that we have StarDump that can operate *not* depending on classes being loaded in that version.

**Comment by per-samuelsson** [comment]
> Well, what if you keep calling MoveNext()? Will it finally pass beyond invalid objects?

I did a quick review of some code, and my guess based on that is that the enumerator will be in some undefined state, but could be worth trying.

**Comment by bigwad** [comment]
>The difference is that we have StarDump that can operate not depending on classes being loaded in that version.

But as I see generic queries over Child table don't work - so the problem is broader than this

**Comment by miyconst** [comment]
>Well, what if you keep calling MoveNext()? Will it finally pass beyond invalid objects?

Did not try that. Going to check whether it works.

**Comment by per-samuelsson** [comment]
> But as I see generic queries over Child table don't work - so the problem is broader than this.

Yes, of course, but to my knowledge, that has always been the case. In 2.x, you can't query on any data where classes are not accessible (i.e. ready to load on demand). I was under the impression that we wouldn't address this in 2.x.

**Comment by miyconst** [comment]
I have checked `DROP TABLE` after a suggestion from @k-rus, and it works. It drops the table and it's data. So, the following works:

```sql
DROP TABLE Child;

SELECT * FROM Parent;
```

This might be the way to go, but it is a dangerous way. There is no guarantee that the table should actually be deleted, because user might have just forgot to load an app for the table.

**Comment by bigwad** [comment]
>Yes, of course, but to my knowledge, that has always been the case. In 2.x, you can't query on any data where classes are not accessible (i.e. ready to load on demand). I was under the impression that we wouldn't address this in 2.x.

Here we can't query event on the base table. I thought this is a bit unexpected.

**Comment by miyconst** [comment]
>I was under the impression that we wouldn't address this in 2.x.

As far as I know this is an issue of the old query processor, and it does not matter on top of which Starcounter version. @k-rus please correct me if I am wrong.

**Comment by k-rus** [comment]
> I was under the impression that we wouldn't address this in 2.x.
> As far as I know this is an issue of the old query processor, and it does not matter on top of which Starcounter version. @k-rus please correct me if I am wrong.

This is why it was written 2.x (not 2.2). New QP comes in 3.x



> Here we can't query event on the base table. I thought this is a bit unexpected.

You can query, but if the result contains an object of child class, it tries to instantiate an object of correct class, i.e., child class, which is not loaded. Instantiating parent class might lead to significant problems with overall system.

**Comment by bigwad** [comment]
> This is why it was written 2.x (not 2.2). New QP comes in 3.x

It's obvious how new QP may help if we do data projection or aggregation, but how it would help with the original query:
```sql
SELECT *
FROM Parent
```

It's a totally host responsibility to handle it.

**Comment by per-samuelsson** [comment]
> It's a totally host responsibility to handle it.

Of course it is. I just said that I knew of this in 2.x and that I was under the impression it was one of those things we would not chase for in 2.x context, but push it to 3.x. Not saying anything if that is the right decision or not (it probably isn't actually).

**Comment by dan31** [comment]
> @dan31, we should have had that vacuum() thingy already in 2.2! 😄

@per-samuelsson:

As genius of enlightenment does prepare any,
So does experience and so genius does,
The One is son of agonizing errors,
The friend of paradoxes's the other one,
And as the god for the contrive is chance,
They all present at once, oh, so many
Miraculous revelations just to us.

**Comment by dan31** [comment]
From David, 15-00 today:

> Ivar recreated all removed classes as dummys and we have now successfully unloaded (a two week old, but should not be a problem) a copy of our intranet db and reloaded it in 2.3! So we will try upgrading the real server in a day or two!

So far, migration goes on! We still need to solve for 2.3+ users how to vacuum, but, let's say, from what I got to know in recent chats, some customers do specifically avoid auto-deletions on migrations, while others do the such auto-deletions quite occasionally. Nevertheless, it doesn't mean that we shouldn't provide a way to vacuum.

**Comment by per-samuelsson** [comment]
> As genius of enlightenment does prepare any,
So does experience and so genius does,
The One is son of agonizing errors,
The friend of paradoxes's the other one,
And as the god for the contrive is chance,
They all present at once, oh, so many
Miraculous revelations just to us.

Anything I would comment on that would just make me look foolish, so I pass with whispered, amazed "oooh". ⚔️

**Comment by k-rus** [comment]
> This is why it was written 2.x (not 2.2). New QP comes in 3.x
> It's obvious how new QP may help if we do data projection or aggregation, but how it would help with the original query:
> 
> SELECT *
> FROM Parent
> It's a totally host responsibility to handle it.

This query returns a projection of all columns of `Parent` table. What will require code host solution is returning objects, i.e.:

```
SELECT p 
FROM Parent p
```
